### PR TITLE
Consumable Adapter - Rewrite to support full ORTB from GO

### PR DESCRIPTION
**Pull request by SuprPhatAnon**
https://github.com/prebid/prebid-server-java/issues/2865

Full rewrite of Adapter and to match the version in Go

### CODE REVIEW

**Review by VeryExtraordinaryUsername — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (VeryExtraordinaryUsername) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:127]: I'm sorry, but this is still not resolved. Please explain to us, why have so differences between Go and Java logic.

Comment (SuprPhatAnon) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:127]: Those changes were not intentional and have been removed.  Thanks for catching it!

**Review by VeryExtraordinaryUsername — APPROVED** [review]

**Comment by VeryExtraordinaryUsername** [comment]
@SuprPhatAnon do you have anything else to add in this PR, or you are done here?

**Comment by SuprPhatAnon** [comment]
@VeryExtraordinaryUsername This PR is complete

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:91]: I suggest merge the App and Site cases into one since the only difference between them is the URL construction, so I believe it'll make the code neater

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:87]: please extract the modifiedRequest, the modified imp and the url into separate variables

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:127]: please order the methods according the code style: the public one first, then all the private ones that are called from the public in the order of calling, then another public one, and then its private ones and so on

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:206]: the builder could be written in a single chain

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:192]: switch is an overkill here, the simple ternary if-operator will be fine here

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:211]: imho an Optional chain will fit here, since there are a lot of not-null checks that could be simplified with the Optional.ofNullable(...)

**Comment by SuprPhatAnon** [comment]
@AntoxaAntoxic Thank you for the feedback.  I can work on getting these changes made.  May I ask what changed since this was previously approved and just need some conflicts resolved before?

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (SuprPhatAnon) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:91]: After looking at the code more deeply there are more that is happening than just the url construction.   The app case extracts the placementId from the Imp and makes an httprequest for each imp instead of combining them as this is how our system works in the backend.  Combining these makes the code way more messy.

**Comment by SuprPhatAnon** [comment]
@AntoxaAntoxic All changes have been made.  Waiting for your approval

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:109]: the update result is never used, shouldn't it be propagated to the imp? 

Also, I think using `UpdateResult` is not necessary, you can just return a new impExt and add it to the imp.

Also looks like this case is not covered by unit tests, so please cover it.

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:212]:
```
 private ExtBidPrebid getExtBidPrebid(Bid bid) {
        return Optional.ofNullable(bid)
                .map(Bid::getExt)
                .map(ext -> parseExtBid(ext))
                .orElse(ExtBidPrebid.builder().build());
 }

 private ExtBidPrebid parseExtBid(ObjectNode bidExt) {
        try {
            return mapper.mapper().treeToValue(bidExt, ExtBidPrebid.class);
        } catch (IllegalArgumentException | JsonProcessingException e) {
            throw new PreBidException(e.getMessage());
        }
    }
```

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:56]: I believe url validation still makes sense

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:139]: the URIs are not checked in the unit tests

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:130]: the result impExt is never checked in the unit tests

File: src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java
Comment (AntoxaAntoxic) [inline src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java:155]: please use `assertThat(result.getErrors()).hasSize(1).first().satisfy(...)`

File: src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java
Comment (AntoxaAntoxic) [inline src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java:195]: `mapper.writeValueAsString()` could be hidden inside the `givenBidResponse`

**Review by AntoxaAntoxic — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (SuprPhatAnon) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java]: This section has been removed as it didn't do anything anyways

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (SuprPhatAnon) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java]: Done

File: src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java
Comment (SuprPhatAnon) [reply src/main/java/org/prebid/server/bidder/consumable/ConsumableBidder.java:108]: Done

File: src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java
Comment (SuprPhatAnon) [reply src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java]: Done

File: src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java
Comment (SuprPhatAnon) [reply src/test/java/org/prebid/server/bidder/consumable/ConsumableBidderTest.java]: Done

**Review by AntoxaAntoxic — APPROVED** [review]
Looks good! Thanks!

**Comment by SuprPhatAnon** [comment]
@AntoxaAntoxic Excellent!  Thanks!  what is the next stop to getting this merged in

**Comment by AntoxaAntoxic** [comment]
It's a merge-freeze stage before the next major release.

**Comment by SuprPhatAnon** [comment]
This update was a simple rebase from the current master branch

**Comment by travisbeale** [comment]
Is there anything else that we need to do in order for this to be merged and included in the next release?  We were approved but not merged into the recent 3.0.0 release.  Thank you.

**Comment by SerhiiNahornyi** [comment]
We tried to not include major bidder changes inside 3.0.

But this PR will definitely be part of upcoming 3.1 release

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**
